tests/utils/test_schemapi.py

```diff
@@ -524,23 +528,23 @@ def chart_error_example__wrong_tooltip_type_in_faceted_chart():
     return (
         alt.Chart(pd.DataFrame({"a": [1]}))
         .mark_point()
-        .encode(tooltip=[{"wrong"}])
+        .encode(tooltip=[{5000}])
         .facet()
     )
 
 
 def chart_error_example__wrong_tooltip_type_in_layered_chart():
     # Error: Wrong data type to pass to tooltip
     return alt.layer(
-        alt.Chart().mark_point().encode(tooltip=[{"wrong"}]),
+        alt.Chart().mark_point().encode(tooltip=[{5000}]),
```

Comment:
These all needed updating since `set` is an `Iterable`.

Reply:
I'm not understanding the reason for updating these. I'm getting the same error message with both `"wrong"` and `5000`. What difference do you see on your end?

```
'['wrong']' is an invalid value for `field`. Valid values are of type 'string' or 'object
```

Reply:
> I'm not understanding the reason for updating these. I'm getting the same error message with both `"wrong"` and `5000`. What difference do you see on your end?
> 
> ```
> '['wrong']' is an invalid value for `field`. Valid values are of type 'string' or 'object
> ```

This one was subtle enough that it took me a while when looking at to notice the problem.

> ```
> '{'wrong'}' is an invalid value for `field`. Valid values are of type 'string' or 'object'.
> ```

```py

original: list[set[str]] = [{"wrong"}]
replaced: list[set[int]] = [{5000}]

```

The original tests relied of `{"wrong"}` being passed through the bottom case of `_to_dict`:
https://github.com/dangotbanned/altair/blob/438ddf994230a547d01349a572404e21e65ae54f/altair/utils/schemapi.py#L533

However, the set elements are now iterated over in:
https://github.com/dangotbanned/altair/blob/438ddf994230a547d01349a572404e21e65ae54f/altair/utils/schemapi.py#L530

---

So this is why the error now reads as a `list`:

> \\[5000\\]

Reply:
@joelostblom I'm not sure how to explain this since my attempts in https://github.com/vega/altair/pull/3501#discussion_r1690163239 and https://github.com/vega/altair/pull/3501#discussion_r1691287871 were not helpful.

I have reverted back to the original tests, to show exactly the failures this resolves - which are in [`bc4f4aa` (#3501)](https://github.com/vega/altair/pull/3501/commits/bc4f4aae84f3b202fb48807397f3b04ee3290d43)

My original commit https://github.com/dangotbanned/altair/commit/1f3dfd33ee273e71c74b67e64ae4410ae0b7c058 fixes the tests that are currently failing

Reply:
I've added some markup to the [CI log](https://github.com/vega/altair/actions/runs/10251914588/job/28361056799?pr=3501) to make it easier to read:


## short test summary info


<details>
<summary>Failed 1</summary>


**FAILED** tests/utils/test_schemapi.py::test_chart_validation_errors[chart_error_example__wrong_tooltip_type_in_faceted_chart-'{'wrong'}' is an invalid value for `field`. Valid values are of type 'string' or 'object'.$] 

```py
AssertionError: Regex pattern did not match.

Regex: "'{'wrong'}' is an invalid value for `field`. Valid values are of type 'string' or 'object'.$"
Input: "'['wrong']' is an invalid value for `field`. Valid values are of type 'string' or 'object'."
```

</details>


<details>
<summary>Failed 2</summary>

**FAILED** tests/utils/test_schemapi.py::test_chart_validation_errors[chart_error_example__wrong_tooltip_type_in_layered_chart-'{'wrong'}' is an invalid value for `field`. Valid values are of type 'string' or 'object'.$]

```py
AssertionError: Regex pattern did not match.

Regex: "'{'wrong'}' is an invalid value for `field`. Valid values are of type 'string' or 'object'.$"
Input: "'['wrong']' is an invalid value for `field`. Valid values are of type 'string' or 'object'."
```

</details>


<details>
<summary>Failed 3</summary>

**FAILED** tests/utils/test_schemapi.py::test_chart_validation_errors[chart_error_example__two_errors_in_layered_chart-Multiple errors were found.\n\nError 1: '{'wrong'}' is an invalid value for `field`. Valid values are of type 'string' or 'object'.\n\nError 2: `Color` has no parameter named 'invalidArgument'\n\n    Existing parameter names are:\n    shorthand      bin         legend   timeUnit   \n    aggregate      condition   scale    title      \n    bandPosition   field       sort     type       \n\n    See the help for `Color` to read the full description of these parameters$]

```py
AssertionError: Regex pattern did not match.

Regex: "Multiple errors were found.\n\nError 1: '{'wrong'}' is an invalid value for `field`. Valid values are of type 'string' or 'object'.\n\nError 2: `Color` has no parameter named 'invalidArgument'\n\n    Existing parameter names are:\n    shorthand      bin         legend   timeUnit   \n    aggregate      condition   scale    title      \n    bandPosition   field       sort     type       \n\n    See the help for `Color` to read the full description of these parameters$"
Input: "Multiple errors were found.\n\nError 1: '['wrong']' is an invalid value for `field`. Valid values are of type 'string' or 'object'.\n\nError 2: `Color` has no parameter named 'invalidArgument'\n\n    Existing parameter names are:\n    shorthand      bin         legend   timeUnit   \n    aggregate      condition   scale    title      \n    bandPosition   field       sort     type       \n\n    See the help for `Color` to read the full description of these parameters"
```

</details>


<details>
<summary>Failed 4</summary>


**FAILED** tests/utils/test_schemapi.py::test_chart_validation_errors[chart_error_example__two_errors_in_complex_concat_layered_chart-Multiple errors were found.\n\nError 1: '{'wrong'}' is an invalid value for `field`. Valid values are of type 'string' or 'object'.\n\nError 2: '4' is an invalid value for `bandPosition`. Valid values are of type 'number'.$]

```py
AssertionError: Regex pattern did not match.

Regex: "Multiple errors were found.\n\nError 1: '{'wrong'}' is an invalid value for `field`. Valid values are of type 'string' or 'object'.\n\nError 2: '4' is an invalid value for `bandPosition`. Valid values are of type 'number'.$"
Input: "Multiple errors were found.\n\nError 1: '['wrong']' is an invalid value for `field`. Valid values are of type 'string' or 'object'.\n\nError 2: '4' is an invalid value for `bandPosition`. Valid values are of type 'number'."
```

</details>


<details>
<summary>Failed 5</summary>


**FAILED** tests/utils/test_schemapi.py::test_chart_validation_errors[chart_error_example__three_errors_in_complex_concat_layered_chart-Multiple errors were found.\n\nError 1: '{'wrong'}' is an invalid value for `field`. Valid values are of type 'string' or 'object'.\n\nError 2: `Color` has no parameter named 'invalidArgument'\n\n    Existing parameter names are:\n    shorthand      bin         legend   timeUnit   \n    aggregate      condition   scale    title      \n    bandPosition   field       sort     type       \n\n    See the help for `Color` to read the full description of these parameters\n\nError 3: '4' is an invalid value for `bandPosition`. Valid values are of type 'number'.$]

```py
AssertionError: Regex pattern did not match.

Regex: "Multiple errors were found.\n\nError 1: '{'wrong'}' is an invalid value for `field`. Valid values are of type 'string' or 'object'.\n\nError 2: `Color` has no parameter named 'invalidArgument'\n\n    Existing parameter names are:\n    shorthand      bin         legend   timeUnit   \n    aggregate      condition   scale    title      \n    bandPosition   field       sort     type       \n\n    See the help for `Color` to read the full description of these parameters\n\nError 3: '4' is an invalid value for `bandPosition`. Valid values are of type 'number'.$"
Input: "Multiple errors were found.\n\nError 1: '['wrong']' is an invalid value for `field`. Valid values are of type 'string' or 'object'.\n\nError 2: `Color` has no parameter named 'invalidArgument'\n\n    Existing parameter names are:\n    shorthand      bin         legend   timeUnit   \n    aggregate      condition   scale    title      \n    bandPosition   field       sort     type       \n\n    See the help for `Color` to read the full description of these parameters\n\nError 3: '4' is an invalid value for `bandPosition`. Valid values are of type 'number'."
```
</details>

### *5 failed*, 1145 passed, 462 skipped, 4 warnings in 42.52s
...

Reply:
Thanks for elaborating further on this and providing the annotated CI logs. In my testing, `list[set[str]]` and `list[set[int]]` both seem to produce the same error message, so what I am confused about is why the fix is to switch from `"wrong"` to `5000` as the set element; it seems like the important part is to change from expecting a set to expecting a list in the error message.

I just pushed commit https://github.com/vega/altair/commit/fb84361ea4142e2a911c416f930c17a240f9c687 to illustrate what I would expect the solution to look like. The tests are passing, so I wonder if there is something I'm missing here that your solution with using an int as the set element accounts for?

Reply:
> I just pushed commit [fb84361](https://github.com/vega/altair/commit/fb84361ea4142e2a911c416f930c17a240f9c687) to illustrate what I would expect the solution to look like. The tests are passing, so I wonder if there is something I'm missing here that your solution with using an int as the set element accounts for?

Thanks @joelostblom that works for me as well and is now in [`8739239` (#3501)](https://github.com/vega/altair/pull/3501/commits/87392396a05faf1039134e47f969d1e47ec19b74)

Apologies I misunderstood you before. 

In https://github.com/vega/altair/pull/3501#discussion_r1691287871 I mentioned that the change was *subtle* - but didn't elaborate that I switched to `{5000}` to draw more attention to it.

Since then - forgetting my own reasoning - I assumed it was a **necessary** change, but it evidently was not 😅 

Appreciate your patience with me also